drivers: led_strip: ws2812_rpi_pico_pio: Add DMA support

TerryGeng opened the pull request:
so that time-consuming I/O takes place in the background. This is especially important when writing to multiple long LED strips.

~~The new Kconfig option that enables this is
`CONFIG_WS2812_STRIP_RPI_PICO_PIO_DMA`.~~ DMA is enabled by the global `CONFIG_DMA` flag and properly populating the `tx` DMA channel in the DT (see below DT example).

This commit mostly mirrors what was done in the pl022 driver and the rpi_pico_pio_spi driver, including the needed DT configuration. If no proper `tx` channel is configured for a device, DMA will not be used for that device (the same behavior as the pl022 driver).

The DT overlay I used when testing:
```
&pio0 {
	status = "okay";

	pio-ws2812 {
		compatible = "worldsemi,ws2812-rpi_pico-pio";
		status = "okay";
		pinctrl-0 = <&ws2812_pio0_default>;
		pinctrl-names = "default";
		bit-waveform = <3>, <3>, <4>;

                /* This device has no tx channel configured so DMA will NOT
                 * be used. */
		ws2812_1: ws2812_1 {
			status = "okay";
			gpios = <&gpio0 23 GPIO_ACTIVE_HIGH>;
			chain-length = <1>;
			color-mapping = <LED_COLOR_ID_GREEN
                                         LED_COLOR_ID_RED
                                         LED_COLOR_ID_BLUE>;
			reset-delay = <280>;
			frequency = <800000>;
		};

                /* This device has proper tx channel configured so DMA will
                 * be used. */
		ws2812_2: ws2812_2 {
			status = "okay";
			gpios = <&gpio0 19 GPIO_ACTIVE_HIGH>;
			chain-length = <64>;
			color-mapping = <LED_COLOR_ID_RED
                                         LED_COLOR_ID_GREEN
                                         LED_COLOR_ID_BLUE>;
			reset-delay = <280>;
			frequency = <800000>;

			dmas = <&dma 1 0 0>;
                        /* DMA slot will be
                         * determined by the driver so what's in here doesn't
                         * matter. */
			dma-names = "tx";
		};
	};
};

&dma {
	status = "okay";
};
```

TerryGeng committed:
drivers: led_strip: ws2812_rpi_pico_pio: Add DMA support

so that time-consuming I/O takes place in the background. This is
especially important when writing to multiple long LED strips.

This commit mostly mirrors what was done in the pl022 driver and the
rpi_pico_pio_spi driver, including the needed DT configuration. DMA will be
used if `CONFIG_DMA` is enabled and proper `tx` channel is defined in the
DT.

The DT overlay I used when testing:
```
&pio0 {
	status = "okay";

	pio-ws2812 {
		compatible = "worldsemi,ws2812-rpi_pico-pio";
		status = "okay";
		pinctrl-0 = <&ws2812_pio0_default>;
		pinctrl-names = "default";
		bit-waveform = <3>, <3>, <4>;

                /* This device has no tx channel configured so DMA will NOT
                 * be used. */
		ws2812_1: ws2812_1 {
			status = "okay";
			gpios = <&gpio0 23 GPIO_ACTIVE_HIGH>;
			chain-length = <1>;
			color-mapping = <LED_COLOR_ID_GREEN
                                         LED_COLOR_ID_RED
                                         LED_COLOR_ID_BLUE>;
			reset-delay = <280>;
			frequency = <800000>;
		};

                /* This device has proper tx channel configured so DMA will
                 * be used. */
		ws2812_2: ws2812_2 {
			status = "okay";
			gpios = <&gpio0 19 GPIO_ACTIVE_HIGH>;
			chain-length = <64>;
			color-mapping = <LED_COLOR_ID_RED
                                         LED_COLOR_ID_GREEN
                                         LED_COLOR_ID_BLUE>;
			reset-delay = <280>;
			frequency = <800000>;

			dmas = <&dma 1 0 0>;
                        /* DMA slot will be
                         * determined by the driver so what's in here
                         * doesn't matter. */
			dma-names = "tx";
		};
	};
};

&dma {
	status = "okay";
};
```

Signed-off-by: Terry Geng <terry@terriex.com>

thedjnK commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -121,3 +121,13 @@ config WS2812_STRIP_RPI_PICO_PIO
+	help
+	  Enable DMA for WS2812 LED strip Raspberry Pi Pico PIO driver. I/O to the LED strip will take place in
+	  the background and the function that updates the pixels will return immediately. Note that this
+	  option will be ignored if no DMA `tx` channel is configured for this device in the devicetree.
```
```suggestion
```
No Kconfig is needed, if the option is there in dts, use DMA for it

ajf58 commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -65,12 +95,130 @@ static int ws2812_led_strip_sm_init(const struct device *dev)
+		 * so this branch will never run in theory.
+		 */
+		return;
+	}
```
This feels like an `ASSERT` would capture this better?

TerryGeng commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -65,12 +95,130 @@ static int ws2812_led_strip_sm_init(const struct device *dev)
+		/* White channel is not supported by LED strip API. */
+		case LED_COLOR_ID_WHITE:
+			color |= 0;
+			break;
```
Done.

TerryGeng commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2023 TOKITA Hiroshi
+ * Copyright (c) 2025 The Zephyr Project Contributors
```
Haha no... I'm too primitive to operate the modern AI machine. Per @kartben I believe this is the preferred copyright notice nowadays.

simonguinot reviewed (commented):
Hi @TerryGeng,

Thanks for this patch. Overall, it seems OK to me. But since I am not familiar with this device, I'll let @soburi handle the review.

Please find a couple of comments below.

TerryGeng commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -80,31 +226,35 @@ static int ws2812_led_strip_update_rgb(const struct device *dev, struct led_rgb
+	/* Wait for the delay needed to latch current color values on
+	 * WS2812 and reset its state machine.
+	 */
+	k_timer_status_sync(&data->reset_on_complete_timer);
```
Thanks for the review. You are correct that this timer (or before this PR, a `k_usleep`) is to wait for the state of WS2812 to reset before updating it again. This reset time, per the datasheet, is several hundred microseconds.

The reason for moving from using `k_usleep` to the timer is: It is unlikely that the LED update function is called twice within the reset time, and calling `k_usleep` (i.e. triggering a context switch) regardless of the actual time between the two function calls to the `update` function is likely wasteful. So it is better to note down the time when the last update takes place and, when on the next call to update, compare the current time to the last update time to decide whether to sleep or not to sleep. I think a timer is a very good tool to guarantee the passage of a certain amount of time (and block the thread only if the wait time hasn't fully elapsed), so I went for it.

On the other hand, `k_work_delayable` is agnostic about when the last update takes place, so I couldn't see how to easily achieve the above.

TerryGeng committed:
drivers: led_strip: ws2812_rpi_pico_pio: Use reset timer instead of sleep

to guarantee two updates to ws2812 happen not faster than the required
reset time of ws2812. Using `k_usleep` forces a context switching, which
might be unnecessary if two time between update calls are not as frequent
as the reset time (which should be the most of the case anyways).

Signed-off-by: Terry Geng <terry@terriex.com>

fabiobaltieri commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -79,32 +244,30 @@ static int ws2812_led_strip_update_rgb(const struct device *dev, struct led_rgb
-				color |= pixels[i].b << (8 * (2 - j));
-				break;
-			}
+	if (!ws2812_led_strip_use_dma(dev)) {
```
swap the condition and keep the non-dma exectuion inline, then on the dma branch do the dma stuff and put a `return 0` at the end, that way you'll avoid that split `} else {`

also my suggestion was to move the whole dma path in its own function so you don't even have a `if defined(CONFIG_DMA)` at all in this function,  so that's still an option, like this could be a

```c
if (ws2812_led_strip_use_dma() {
   return ws2812_led_strip_start_dma_put();
}

<normal non dma code>
```

and then those two  functions could be no-op in the non dma case, but I think


```c
#if defined(CONFIG_DMA)
if (ws2812_led_strip_use_dma() {
   <do the dma thing>
   return 0;
}
#endif
```

is also valid, if you go with that you can remove the condition from`ws2812_led_strip_use_dma` and just undefine it in the non-dma case

am I making sense?

TerryGeng replied:
If you want me to populate the pixel buf in that `ws2812_led_strip_start_dma_put` function, I can also do that.

fabiobaltieri reviewed (approved):
good stuff

TerryGeng commented:
@fabiobaltieri Thanks for your rapid response and patience!

simonguinot commented:
@soburi is that good for you ?

simonguinot reviewed (approved):
Thanks @TerryGeng !
